### DISCUSSION TITLE
How to "cancel" a body using `request`?

### DISCUSSION THREAD

**Discussion by WTCT-TOP** [nodejs/undici#3914] [Q&A] [answered]
Hi,

In [this section of the README](https://github.com/nodejs/undici#garbage-collection) it states that "it is important to always either consume or cancel the response body", but only shows examples with `fetch` (and not `request`)..

1. Does this apply to `request` too? (from the manual tests I did it seems like it, but just to make sure..)
2. Is `response.body.destroy()` the correct way to "cancel" the body? If it is, how can I make sure it will not throw/emit an error?

Thanks

**Comment by metcoder95** [comment, marked as answer]
1. Yes, it also applies.
2. Using `response.body.dump()` will be better.

**Reply by WTCT-TOP** [reply]
Thank you for the quick response! :)

1. its better to use dump even if I don't care about the body at all? there is no way to just destroy the socket at that point (am I missing something about how HTTP works?)?

```javascript
const response = await request('...');
if (response.statusCode === 401) {
  // I don't care about the body at all here
  throw new Error('401');
}
```

2. dump has a `limit` option which defaults to `131072`, what if the body is bigger than that?

3. should I update the docs?

**Comment by metcoder95** [comment]
1. Yes. Even if you don't care about it, the response is still buffered in memory for consumption. Only after the `response` object is collected by the GC, the memory of the response's body will be eventually freed. Under high-traffic situations, this can cause some harm to the overall performance of your app.
2. The `dump` method will discard the body up to that threshold to re-use the same connection. If above the threshold, the connection will be dropped (and eventually a new one will be made).
3. Sure! Any contribution is really welcomed!

**Reply by WTCT-TOP** [reply]
1. So the main reason to `dump` is to be able to reuse connections?
2. This is the current code in the docs, which unlike `dump` does not have a limit (I think), which one is "wrong"?
```javascript
for await (const chunk of res.body) {
  // force consumption of body
}
```
3. Does this makes sense as an example for request?
```javascript
// Do
const response = await request(url)
  .then(res => {
    response.body.dump().catch(err => {
      console.error('Error while dumping the body');
      // is there anything I can/should do here?
    });
    return res.headers;
  });

// Do not
const response = await request(url)
  .then(res => res.headers);
```

**Reply by metcoder95** [reply]
1. Partially, also to help dump the response's data if not needed at all
2. A `res.body` can be added as option
3. Yes, sure; I'd just simplify it to:

```js
// Do
const response = await request(url)
const headers = request.headers // Do something with headers
await response.body.dump()

// Do not
const response = await request(url)
const headers = request.headers // Do something with headers
```

Something down the line

**Reply by weyert** [reply]
Maybe worth having an eslint plugin for this to warn people of the potential memory leak?

**Reply by metcoder95** [reply]
Something like that will be really welcomed
